pkg/pillar/cmd/zedrouter/zedrouter.go

```diff
@@ -1224,58 +1241,22 @@ func findBridge(bridgeName string) (*netlink.Bridge, error) {
 
 // XXX Need additional logic for IPv6 underlays.
 func getUlAddrs(ctx *zedrouterContext,
-	ifnum int, appNum int,
-	status *types.UnderlayNetworkStatus,
-	netInstStatus *types.NetworkInstanceStatus) (string, string, error) {
-
-	log.Functionf("getUlAddrs(%d/%d)\n", ifnum, appNum)
-
-	bridgeIPAddr := ""
-	appIPAddr := ""
-
-	// Allocate bridgeIPAddr based on BridgeMac
-	log.Functionf("getUlAddrs(%d/%d for %s) bridgeMac %s\n",
-		ifnum, appNum, netInstStatus.UUID.String(),
-		status.BridgeMac.String())
-	var err error
-	var addr string
-	addr, err = lookupOrAllocateIPv4(ctx, netInstStatus,
-		status.BridgeMac)
-	if err != nil {
-		log.Errorf("getUlAddrs: Bridge IP address allocation failed %s\n", err)
-		return bridgeIPAddr, appIPAddr, err
-	} else {
-		bridgeIPAddr = addr
-	}
-
-	if status.AppIPAddr != nil {
-		// Static IP assignment case.
-		// Note that appIPAddr can be in a different subnet.
-		// Assumption is that the config specifies a gateway/router
-		// in the same subnet as the static address.
-		appIPAddr = status.AppIPAddr.String()
-		recordIPAssignment(ctx, netInstStatus, status.AppIPAddr,
-			status.Mac)
-	} else if status.Mac != "" {
-		// XXX or change type of VifInfo.Mac to avoid parsing?
-		var mac net.HardwareAddr
-		mac, err = net.ParseMAC(status.Mac)
-		if err != nil {
-			log.Fatal("ParseMAC failed: ", status.Mac, err)
-		}
-		log.Functionf("getUlAddrs(%d/%d for %s) app Mac %s\n",
-			ifnum, appNum, netInstStatus.UUID.String(), mac.String())
-		addr, err = lookupOrAllocateIPv4(ctx, netInstStatus, mac)
-		if err != nil {
-			log.Errorf("getUlAddrs: App IP address allocation failed: %s\n", err)
-			return bridgeIPAddr, appIPAddr, err
-		} else {
-			appIPAddr = addr
-		}
+	ifnum int, appNum int, appID string, status *types.UnderlayNetworkStatus,
+	netInstStatus *types.NetworkInstanceStatus) (string, error) {
+	log.Functionf("getUlAddrs(%d/%d) for %s\n", ifnum, appNum, appID)
+	networkID := netInstStatus.UUID.String()
+	status.BridgeIPAddr = netInstStatus.BridgeIPAddr
+	ipAddr, err := networkInstanceAppIPAddrGet(ctx, appID,
```

Comment:
You seem to have lost the case of a static IP here; the check status.AppIPAddr != "" in the original code.
In that case you'd still have an appNumOnNet number (picked when the UlConfig is processed) but not use it for allocation of the IP address.